File: python/day04.py
import re

def parse(line):
    parts = re.split("-|,", line)
    return [int(p) for p in parts]

def contains(x, y, a, b):
    return x >= a and y <= b

def overlaps(x, y, a, b):
    return (x <= b and y >= a) or (a <= y and b >= x)

def solution():
    data = [parse(s.strip()) for s in open("../data/04.txt").readlines()]
    res1 = sum(contains(s[0], s[1], s[2], s[3]) or
            contains(s[2], s[3], s[0], s[1]) for s in data)
    res2 = sum(overlaps(s[0], s[1], s[2], s[3]) for s in data)

    print(f'Answer 1: {res1}\nAnswer 2: {res2}')

File: rust/src/day04.rs
use crate::common;
use regex::Regex;

fn parse(s: &str) -> Vec<i64> {
    lazy_static! {
        static ref REGEX: Regex = Regex::new(r"-|,").unwrap();
    }
    REGEX
        .split(s)
        .map(|val| val.parse::<i64>().unwrap())
        .collect()
}

fn contains(x: i64, y: i64, a: i64, b: i64) -> bool {
    x >= a && y <= b
}

fn overlaps(x: i64, y: i64, a: i64, b: i64) -> bool {
    (x <= b && y >= a) || (a <= y && b >= x)
}

pub(crate) fn solution() {
    let data: Vec<Vec<i64>> = common::read_lines(&common::data_file(4))
        .unwrap()
        .iter()
        .map(|s| parse(s))
        .collect();

    let res1: i64 = data
        .iter()
        .map(|s| (contains(s[0], s[1], s[2], s[3]) || contains(s[2], s[3], s[0], s[1])) as i64)
        .sum();
    let res2: i64 = data
        .iter()
        .map(|s| overlaps(s[0], s[1], s[2], s[3]) as i64)
        .sum();
    println!("Answer 1: {}, Answer 2: {}", res1, res2);
}
